### ISSUE TITLE
Incorporating my LM into the CTC-beam search

### ISSUE THREAD

**Issue by alanbekker** [mozilla/DeepSpeech#720]
Hi guys,

Do you know how can incorporate my LM into the CTC beam-search decoding?

**Comment by kdavis-mozilla** [comment]
I can't cover all the details here. But basically there are at least two ways

1. Follow the lead of Deep Speech[[1](https://arxiv.org/abs/1412.5567)] and Deep Speech 2[[2](https://arxiv.org/abs/1512.02595)]
2. Integrate the language model into Tensorflow itself, implementing a custom `BeamScorer`[[3](https://github.com/tensorflow/tensorflow/blob/198f9a16ebf2c9ff2e842f1cce99ddf02c7a9849/tensorflow/core/util/ctc/ctc_beam_search.h#L77)] that includes a LM

**Comment by alanbekker** [comment]
Thanks for the answer.

You mean to integrate the two options into one solution, right?
 In the tensorflow CTC_decoder I have no access to  the beam search
decoding process and therefore you suggest to integrate the LM(as DS1 and
DS2 did)  into my own BeamScorer function in C++ and then recompile
tensotflow?

**Comment by kdavis-mozilla** [comment]
I mean the two options are either/or, in other words one would implement one, but not the other.

If you compile Tensorflow from source, you have access to the CTC decoder. So you could modify the CTC decoder. In fact Tensorflow supplies hooks to do just this. This is the second option above.

The first option requires no modification of Tensorflow. One simply uses the CTC decoder results and does not need to modify Tensorflow.

**Comment by alanbekker** [comment]
My question  was that the  first option you suggested is impossible to
implement without the second (maybe I have some mis-understanding :(  and
be glad if you correct me ) since in order to beam search the following
function :
 [image: Inline image 1]
I need to be able to track the probabilities of each character in the
decoding...but currently tf api provides the likelihood of the whole
decoded sentence.

**Comment by kdavis-mozilla** [comment]
You can implement the first without doing the second or the second without doing the first.

The C++ Tensorflow API provides for integrating language models into the CTC decoder. See for example[[1](https://github.com/tensorflow/tensorflow/blob/198f9a16ebf2c9ff2e842f1cce99ddf02c7a9849/tensorflow/core/util/ctc/ctc_beam_search.h#L77)]. The python API, however, does not allow for such tight language model integration. One has to do the work described in Deep Speech and Deep Speech 2 "by-hand".

**Comment by alanbekker** [comment]
Thanks again for your answer,
I'm sorry, but I think you are wrong.  Doing it "by hand" would not be
optimal as using the  c++ ctc_decoder.
The reason is within the c++ api ctc-decoder you have access to the softmax
probabilities for each character in the sequence while using the python api
you have access only the the top N paths of the decoder (causing the beam
search in the first option being much optimal than the second)

**Comment by kdavis-mozilla** [comment]
@alanbekker I never mentioned one way or the other as being "optimal," they're simply different.

However, for a project such as this, one has to realize that not every user will want to use our modified version of TensorFlow. So creating a custom `BeamScorer` immediately turns away potential users.

**Comment by alanbekker** [comment]
Absolutely agree, thanks again!
